src/backend/distributed/commands/schema_based_sharding.c

```diff
@@ -191,13 +285,192 @@ ErrorIfIllegalPartitioningInTenantSchema(Oid parentRelationId, Oid partitionRela
 							   "supported when the parent and the child "
 							   "are in different schemas")));
 	}
 }
 
 
+/*
+ * CreateTenantSchemaColocationId returns new colocation id for a tenant schema.
+ */
+uint32
+CreateTenantSchemaColocationId(void)
+{
+	int shardCount = 1;
+	int replicationFactor = 1;
+	Oid distributionColumnType = InvalidOid;
+	Oid distributionColumnCollation = InvalidOid;
+	uint32 schemaColocationId = CreateColocationGroup(
+		shardCount, replicationFactor, distributionColumnType,
+		distributionColumnCollation);
+	return schemaColocationId;
+}
+
+
+/*
+ * GetRegularAndForeignCitusTablesInSchema returns all nonshard relation ids,
+ * with relkind = relation or partitioned or foreign, inside given schema.
+ */
+static List *
+GetRegularAndForeignCitusTablesInSchema(Oid schemaId)
+{
+	List *relationIdList = NIL;
+
+	/* scan all relations in pg_class and return all tables inside given schema */
+	Relation relationRelation = relation_open(RelationRelationId, RowExclusiveLock);
+
+	ScanKeyData scanKey[1] = { 0 };
+	ScanKeyInit(&scanKey[0], Anum_pg_class_relnamespace, BTEqualStrategyNumber,
+				F_OIDEQ, ObjectIdGetDatum(schemaId));
+	SysScanDesc scanDescriptor = systable_beginscan(relationRelation, ClassNameNspIndexId,
+													true, NULL, 1, scanKey);
+
+	HeapTuple heapTuple = NULL;
+	while (HeapTupleIsValid(heapTuple = systable_getnext(scanDescriptor)))
+	{
+		Form_pg_class relationForm = (Form_pg_class) GETSTRUCT(heapTuple);
+		char *relationName = NameStr(relationForm->relname);
+		Oid relationId = get_relname_relid(relationName, schemaId);
+
+		if (!OidIsValid(relationId))
+		{
+			ereport(ERROR, errmsg("table %s is dropped by a concurrent operation",
+								  relationName));
+		}
+
+		/* skip shards */
+		if (RelationIsAKnownShard(relationId))
+		{
+			continue;
+		}
+
+		if (RegularTable(relationId) || IsForeignTable(relationId))
+		{
+			relationIdList = lappend_oid(relationIdList, relationId);
+		}
+	}
+
+	systable_endscan(scanDescriptor);
+	relation_close(relationRelation, RowExclusiveLock);
+
+	return relationIdList;
+}
+
+
+/*
+ * EnsureSchemaCanBeDistributed ensures the schema can be distributed.
+ * Caller should take required the lock on relations and the schema.
+ *
+ * It checks:
+ *  - Schema name is in the allowed-list,
+ *  - Schema is not owned by an extension,
+ *  - Current user should be the owner of tables under the schema,
+ *  - Table kinds are supported,
+ *  - Only Citus local and Postgres local tables exist under the schema.
+ */
+static void
+EnsureSchemaCanBeDistributed(Oid schemaId, List *schemaTableIdList)
+{
+	/* Ensure schema name is allowed */
+	EnsureTenantSchemaNameAllowed(schemaId);
+
+	/* Any schema owned by extension is not allowed */
+	ObjectAddress *schemaAddress = palloc0(sizeof(ObjectAddress));
+	ObjectAddressSet(*schemaAddress, NamespaceRelationId, schemaId);
+	if (IsAnyObjectAddressOwnedByExtension(list_make1(schemaAddress), NULL))
```

Comment:
if I do CREATE EXTENSION postgis IN SCHEMA foo; is foo then covered by this check, or is it only covered when the extension script does a CREATE SCHEMA foo?

Reply:
It is only covered if extension script created a schema now.

Reply:
Then we should also cover the other way around. And we also disallow if any object in the schema is owned by an extension?

Reply:
>  we also disallow if any object in the schema is owned by an extension?

(I don't think we need to do that)